### PR TITLE
Add support for weights in both explicit and implicit models

### PR DESCRIPTION
**Pull request by lindauruchurtu**
This PR adds weight support to explicit and implicit models, by modifying the losses and the fit methods in the corresponding class.

TODO: Sample dataset for test.

### CODE REVIEW

File: spotlight/losses.py [hunk 1]
```diff
@@ -79,8 +83,8 @@ def bpr_loss(positive_predictions, negative_predictions, mask=None):
        uncertainty in artificial intelligence. AUAI Press, 2009.
     """
 
-    loss = (1.0 - F.sigmoid(positive_predictions -
-                            negative_predictions))
+    loss = (1.0 - F.sigmoid((positive_predictions -
+                             negative_predictions) * weights))
```
Comment (maciejkula) [inline spotlight/losses.py]: I think the `weights` term should be outside the `(1.0 - F.sigmoid(...))` term. I don't think we want to ever allow this term to become negative.

File: spotlight/losses.py [hunk 2]
```diff
@@ -112,8 +118,8 @@ def hinge_loss(positive_predictions, negative_predictions, mask=None):
         The mean value of the loss function.
     """
 
-    loss = torch.clamp(negative_predictions -
-                       positive_predictions +
+    loss = torch.clamp((negative_predictions -
+                        positive_predictions) * weights +
```
Comment (maciejkula) [inline spotlight/losses.py]: Again, I think the `weights` term should be outside the outermost parenthesis.

So it should be equivalent to
```python
loss = torch.clamp(negative_predictions - positive_predictions + 1.0), 0.0)
loss *= weights
```

File: spotlight/losses.py [hunk 3]
```diff
@@ -188,10 +203,12 @@ def regression_loss(observed_ratings, predicted_ratings):
 
     assert_no_grad(observed_ratings)
 
-    return ((observed_ratings - predicted_ratings) ** 2).mean()
+    return (
+        ((observed_ratings - predicted_ratings) * weights) ** 2
```
Comment (maciejkula) [inline spotlight/losses.py]: I don't think we should be squaring the weights: compute the squared error, _then_ weight.

File: spotlight/losses.py [hunk 4]
```diff
@@ -212,4 +231,4 @@ def poisson_loss(observed_ratings, predicted_ratings):
 
     assert_no_grad(observed_ratings)
 
-    return (predicted_ratings - observed_ratings * torch.log(predicted_ratings)).mean()
+    return ((predicted_ratings - observed_ratings) * weights * torch.log(predicted_ratings)).mean()
```
Comment (maciejkula) [inline spotlight/losses.py]: Ditto, it's just a scaling term before the mean.

File: tests/factorization/test_implicit.py [hunk 5]
```diff
@@ -23,6 +23,39 @@ def _min_max_scale(arr):
     return (arr - arr_min) / (arr_max - arr_min)
 
 
+@pytest.mark.parametrize('weights, loss, expected', [
```
Comment (maciejkula) [inline tests/factorization/test_implicit.py]: Two comments:
1. Can we instantiate the weights arrays inside the function? You could use one of `(0, 1)` as a scaling factor in the pytest args, and do the multiplication (with `np.ones`) inside the test function. I think that would make it cleaner.
2. These tests will run forever :) Maybe we could run them on a smaller subset of the data?

Comment (lindauruchurtu) [inline tests/factorization/test_implicit.py]: I'll leave point 2 for tomorrow.

**Comment by maciejkula** [comment]
:+1:  great, thanks!
